internal/server/ingest/ingest.go

```diff
@@ -0,0 +1,229 @@
+package ingest
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"log/slog"
+	"path/filepath"
+	"sync"
+	"time"
+
+	"github.com/ubuntu/ubuntu-insights/internal/server/ingest/database"
+	"github.com/ubuntu/ubuntu-insights/internal/server/ingest/models"
+	"github.com/ubuntu/ubuntu-insights/internal/server/ingest/processor"
+)
+
+// Service represents the ingest service that processes and uploads reports to the database.
+type Service struct {
+	cm dConfigManager
+	db dbManager
+
+	// This context is used to interrupt any action.
+	// It must be the parent of gracefulCtx.
+	ctx    context.Context
+	cancel context.CancelFunc
+
+	// This context waits until the next blocking Recv to interrupt.
+	gracefulCtx    context.Context
+	gracefulCancel context.CancelFunc
+
+	mu      sync.Mutex
+	workers map[string]context.CancelFunc
+}
+
+type dbManager interface {
+	Upload(ctx context.Context, app string, data *models.TargetModel) error
+	Close() error
+}
+
+type dConfigManager interface {
+	Load() error
+	Watch(context.Context) (<-chan struct{}, <-chan error, error)
+	AllowList() []string
+	BaseDir() string
+}
+
+type options struct {
+	dbConnect func(ctx context.Context, cfg database.Config) (dbManager, error)
+}
+
+// Options is a function that modifies the options for the ingest service.
+type Options func(*options)
+
+// New creates a new ingest service with the provided database manager and connects to the database.
+func New(ctx context.Context, cm dConfigManager, dbConfig database.Config, args ...Options) (*Service, error) {
+	opts := options{
+		dbConnect: func(ctx context.Context, cfg database.Config) (dbManager, error) {
+			return database.Connect(ctx, cfg)
+		},
+	}
+
+	for _, opt := range args {
+		opt(&opts)
+	}
+
+	if err := cm.Load(); err != nil {
+		return nil, fmt.Errorf("failed to load configuration: %v", err)
+	}
+
+	dctx, dcancel := context.WithTimeout(ctx, 10)
+	defer dcancel()
+	db, err := opts.dbConnect(dctx, dbConfig)
+	if err != nil {
+		return nil, fmt.Errorf("failed to connect to database: %v", err)
+	}
+
+	ctx, cancel := context.WithCancel(ctx)
+	gCtx, gCancel := context.WithCancel(ctx)
+
+	return &Service{
+		cm:             cm,
+		db:             db,
+		ctx:            ctx,
+		cancel:         cancel,
+		gracefulCtx:    gCtx,
+		gracefulCancel: gCancel,
+
+		mu:      sync.Mutex{},
+		workers: make(map[string]context.CancelFunc),
+	}, nil
+}
+
+// Run starts the ingest service.
+//
+// It watches the configured location for new reports that have been saved to disk.
+// It will then process, validate, and then upload the reports to the database.
+func (s *Service) Run() error {
+	slog.Info("Ingest service started")
+
+	select {
+	case <-s.gracefulCtx.Done():
+		return errors.New("server is already shutting down")
+	default:
+	}
+
+	reloadEventCh, cfgWatchErrCh, err := s.cm.Watch(s.gracefulCtx)
+	if err != nil {
+		return fmt.Errorf("failed to start watch configuration: %v", err)
+	}
+
+	// 2) initial sync
```

Comment:
This seems like a friendly step-by-step explanation of something but only step 2 remained as numbered. 
```suggestion
	// Initial sync
```

otherwise identify the other numbered steps :)